provider/consumer.py

```diff
@@ -96,32 +99,33 @@ def run(self):
 
                 payload = {}
                 payload['messages'] = messages
-                logging.info("Firing trigger {} with {} messages with a total size of {} bytes".format(self.trigger, len(messages), messageSize))
-
-                try:
-                    response = requests.post(self.triggerURL, json=payload)
-                    status_code = response.status_code
-                    logging.info("Repsonse status code {}".format(status_code))
-
-                    if status_code >= 400 and status_code < 500:
-                        # TODO need to kill this trigger
-                        # However, 429 is the status code for a throttled response
-                        logging.info('Need to kill this trigger due to status code {}'.format(status_code))
-                    elif status_code >= 500:
-                        # TODO OW server error... now what?
-                        logging.error('Error talking to OW, status code {}'.format(status_code))
-                    else:
-                        # manually commit offsets only upon successfully
-                        # firing the trigger
-                        self.consumer.commit()
-                except requests.exceptions.RequestException as e:
-                    # TODO network issue talking to OW... now what?
-                    logging.error('Error talking to OW: {}'.format(e))
-
-                # TODO
-                # error handling
-                # retries? I suppose certain status codes warrant it...
-                # only commit offset if post was successful?
+                retry = True
+
+                logging.info("Firing trigger {} with {} messages with a total size of {} bytes".format(self.trigger,
+                   len(messages), messageSize))
+
+                while retry:
+                    try:
+                        response = requests.post(self.triggerURL, json=payload)
+                        status_code = response.status_code
+                        logging.info("Repsonse status code {}".format(status_code))
+
+                        # Manually commit offset if the trigger was fired successfully. Retry firing the trigger if the
+                        # service was unavailable (503), an internal server error occurred (500), the request timed out
+                        # (408), or the request was throttled (429).
+                        if status_code == 200:
+                            self.consumer.commit()
+                            retry = False
+                        elif status_code != 503 and status_code != 500 and status_code != 408 and status_code != 429:
+                            logging.error('Error talking to OpenWhisk, status code {}'.format(status_code))
+                            retry = False
+                            self.shouldRun = False
```

Comment:
regarding the quota -- the situation I have in mind is a user who has not put a credit card on file and has exhausted their monthly free tier.     All  further usage (in the month) will be rejected.   Actually I don't know if we decided what code they get -- it may not be 429.   Anyway, in this case, we don't want to retry.

Regarding bounded retries -- I'm just scared of things running out of control if there's a bug anywhere.     Safety first, right?   I don't want a bug or failure to result in a DoS attack -- I'd rather drop events.   Note that I don't intend to "drop the feed", just drop events for the duration of the error condition rather than having huge amounts of work back up.

Another question: what if the kafka feed is just too heavy for OW to ever catch up?   e.g. if the kafka feed is generating 1 million messages per second and OW can consume only 1000, we don't want millions or billions of messages to back up, I think.